projects/serializers/project.py

```diff
@@ -0,0 +1,124 @@
+import copy
+from typing import List, NamedTuple, Type
+
+from django.db import transaction
+from rest_framework import serializers
+from rest_framework.exceptions import ValidationError
+from rest_framework.serializers import Serializer
+
+from projects.models import Project, ProjectPhase, ProjectPhaseSection, ProjectType
+from projects.serializers.section import create_section_serializer
+
+
+class ProjectTypeSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Project
+        fields = ["name"]
+
+
+class SectionData(NamedTuple):
+    section: ProjectPhaseSection
+    serializer_class: Type[Serializer]
+
+
+class ProjectSerializer(serializers.ModelSerializer):
+    class Meta:
+        model = Project
+        fields = [
+            "user",
+            "created_at",
+            "modified_at",
+            "name",
+            "identifier",
+            "type",
+            "attribute_data",
+            "phase",
+            "geometry",
+            "id",
+        ]
+        read_only_fields = ["phase", "type", "created_at", "modified_at"]
+
+    def should_validate_attributes(self):
+        validate_field_data = self.context["request"].data.get(
+            "validate_attribute_data", False
+        )
+        return serializers.BooleanField().to_internal_value(validate_field_data)
+
+    def generate_sections_data(self, phase: ProjectPhase) -> List[SectionData]:
+        sections = []
+        for section in phase.sections.order_by("index"):
+            serializer_class = create_section_serializer(section)
+            section_data = SectionData(section, serializer_class)
+            sections.append(section_data)
+
+        return sections
+
+    def validate_attribute_data(self, attribute_data):
+        # Get instance phase or use the first phase for the project type
+        phase = (
+            self.instance.phase
+            if self.instance and getattr(self.instance, "phase", None)
+            else ProjectPhase.objects.get(project_type__name="asemakaava", index=0)
```

Comment:
Maybe `ProjectPhase.objects.get(project_type=self.instance.type, index=0)` could be a more suitable filter.
Could even be `filter(...).first()` instead of `get(...)` since there's no unique constraints.

Reply:
Ah, wasn't able to read all the lines in this assignment, please ignore my last comment. :laughing: